File: app/src/test/java/com/example/mike/mp3player/client/views/viewholders/MySongViewHolderTest.java
package com.example.mike.mp3player.client.views.viewholders;

import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaBrowserCompat.MediaItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.utils.TimerUtils;
import com.example.mike.mp3player.commons.MediaItemBuilder;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class MySongViewHolderTest extends MediaItemViewHolderTestBase {

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        when(view.findViewById(R.id.artist)).thenReturn(mock(TextView.class));
        when(view.findViewById(R.id.title)).thenReturn(mock(TextView.class));
        when(view.findViewById(R.id.duration)).thenReturn(mock(TextView.class));
        when(view.findViewById(R.id.song_item_album_art)).thenReturn(mock(ImageView.class));
        this.mediaItemViewHolder = spy(new MySongViewHolder(view, albumArtPainter));
    }

    @Test
    public void testBindMediaItem() {

        final String expectedTitle = "title";
        final String expectedArtist = "artist";
        final long originalDuration = 34978L;
        final String expectedDuration = TimerUtils.formatTime(originalDuration);
        MediaItem mediaItem = new MediaItemBuilder("id")
                .setTitle(expectedTitle)
                .setArtist(expectedArtist)
                .setDuration(originalDuration)
                .build();
        this.mediaItemViewHolder.bindMediaItem(mediaItem);
        MySongViewHolder songViewHolder = (MySongViewHolder) mediaItemViewHolder;
        verify(songViewHolder.getTitle(), times(1)).setText(expectedTitle);
    }
}

File: client/src/testUnitTest/java/com/github/goldy1992/mp3player/client/views/viewholders/MySongViewHolderTest.kt
package com.github.goldy1992.mp3player.client.views.viewholders

import android.net.Uri
import android.widget.ImageView
import android.widget.TextView
import com.github.goldy1992.mp3player.client.databinding.SongItemMenuBinding
import com.github.goldy1992.mp3player.client.databinding.ViewHolderMediaPlayerBinding
import com.github.goldy1992.mp3player.client.utils.TimerUtils.formatTime
import com.github.goldy1992.mp3player.commons.Constants
import com.github.goldy1992.mp3player.commons.MediaItemBuilder
import com.nhaarman.mockitokotlin2.*
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class MySongViewHolderTest : MediaItemViewHolderTestBase<MySongViewHolder>() {
    lateinit var view : SongItemMenuBinding

    @Before
    override fun setup() {
        super.setup()
        scenario.onActivity {
            this.view = SongItemMenuBinding.inflate(it.layoutInflater)
            this.mediaItemViewHolder = MySongViewHolder(view, albumArtPainter)
        }
    }

    @Test
    fun testBindMediaItemValidTitle() {
        val expectedTitle = "title"
        val mediaItem = MediaItemBuilder("id")
                .setTitle(expectedTitle)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedTitle, view.title.text)
    }

    @Test
    fun testBindMediaItemNullTitleNoFileName() {
        val expectedTitle = Constants.UNKNOWN
        val mediaItem = MediaItemBuilder("id")
                .setTitle(null)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedTitle, view.title.text)
    }

    @Test
    fun testExtractTitleNullTitleFileNameNoExtension() {
        val expectedTitle = "no_extension"
        val mediaItem = MediaItemBuilder("id")
                .setTitle(null)
                .setFileName(expectedTitle)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedTitle, view.title.text)
    }

    @Test
    fun testExtractTitleNullTitleFileNameWithExtension() {
        val expectedTitle = "file_with_extension"
        val fileName = "$expectedTitle.mp3"
        val mediaItem = MediaItemBuilder("id")
                .setTitle(null)
                .setFileName(fileName)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedTitle, view.title.text)
    }

    @Test
    fun testSetAlbumArt() {
        val expectedUri = mock<Uri>()
        val mediaItem = MediaItemBuilder("id")
                .setAlbumArtUri(expectedUri)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        verify(mediaItemViewHolder.albumArtPainter, times(1))!!
                .paintOnView(view.albumArt, expectedUri)
    }

    @Test
    fun testSetArtist() {
        val expectedArtist = "artist"
        val mediaItem = MediaItemBuilder("id")
                .setArtist(expectedArtist)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedArtist, view.artist.text)
    }

    @Test
    fun testSetArtistNull() {
        val expectedArtist = Constants.UNKNOWN
        val mediaItem = MediaItemBuilder("id")
                .setArtist(null)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedArtist, view.artist.text)
    }

    @Test
    fun testSetDuration() {
        val originalDuration = 34978L
        val expectedDuration = formatTime(originalDuration)
        val mediaItem = MediaItemBuilder("id")
                .setDuration(originalDuration)
                .build()
        mediaItemViewHolder.bindMediaItem(mediaItem)
        assertEquals(expectedDuration, view.duration.text)
    }
}